### DISCUSSION TITLE
Writing a generic function that can return type safe masks.

### DISCUSSION THREAD

**Discussion by WesleyYue** [colinhacks/zod#1814] [Q&A] [answered]
I'm pretty deep into the weeds with type script generics, but this is what I have so far:

```ts
import { z } from "zod";

const EnvVarRaw = {
  ENV_VAR_1: z.string(),
  ENV_VAR_2: z.string(),
  ENV_VAR_3: z.string(),
};

const EnvVarSchema = z.object(EnvVarRaw);

export function getEnvVar<
  Mask extends {
    [k in keyof T]?: true;
  },
  Schema extends z.ZodObject<T>,
  T extends z.ZodRawShape
>(
  mask: Mask,
  schema: Schema
): z.objectOutputType<Pick<T, Extract<keyof T, keyof Mask>>, z.ZodTypeAny> {
  const secrets = schema.pick(mask).parse(process.env);

  return secrets;
}

getEnvVar(
  {
    ENV_VAR_1: true,
  },
  EnvVarSchema
);

```

The goal is to create a getEnvVar function that doesn't require putting in the EnvVarSchema as a parameter. I'm stuck here. What typescript magic do I have to conjure to make this possible?

I can of course do `schema.pick(mask).parse(process.env)` directly and that will get me the exact type I want (properly masked return type), though I'm curious as to how to achieve the above as a learning exercise.

**Comment by JacobWeisenburger** [comment]
I don't see the function definition for `getEnvVarAny`. Is that a typo?

**Reply by WesleyYue** [reply]
Sorry, I forgot to specify. You can assume getEnvVarAny() returns `{ENV_VAR_1: any, ENV_VAR_2: any, ENV_VAR_3: any}`. I should have just used `process.env` in the example

**Reply by JacobWeisenburger** [reply]
Shouldn't it be this?
```ts
{ENV_VAR_1: string, ENV_VAR_2: string, ENV_VAR_3: string}
```

**Reply by WesleyYue** [reply]
Do you mean `getEnvVarAny()` should have return type of `{ENV_VAR_1: string, ENV_VAR_2: string, ENV_VAR_3: string}`? In that case, no, and the goal is to use zod to validate that the envVar's exist and is type `string`. I think the example makes more sense if you forget about `getEnvVarAny` and substitute `process.env` instead, sorry.

**Reply by WesleyYue** [reply]
The goal is to use zod to validate that the envVar's exist and is type string, _and allow the caller of `getEnvVar()` to specify and pick off only the env vars that they need to get a return type that only consists of the masked env vars._

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const envVarSchema = z.object( {
    ENV_VAR_1: z.string(),
    ENV_VAR_2: z.string(),
    ENV_VAR_3: z.string(),
} )

/* for node use: process.env */
/* for deno use: Deno.env.toObject() */
const env = {
    ENV_VAR_1: 'rick',
    ENV_VAR_2: 'morty',
    ENV_VAR_3: 'mr meeseeks',
}

export function getEnvVar<
    Shape extends z.ZodRawShape,
    Mask extends { [ k in keyof Shape ]?: true },
> ( mask: Mask, schema: z.ZodObject<Shape> ) {
    return schema.pick( mask ).parse( env )
}

const result1 = getEnvVar( { ENV_VAR_1: true }, envVarSchema )
// const result1: { ENV_VAR_1: string }
console.log( result1 ) // { ENV_VAR_1: 'rick' }

const result2 = getEnvVar( { ENV_VAR_2: true }, envVarSchema )
// const result2: { ENV_VAR_2: string }
console.log( result2 ) // { ENV_VAR_2: 'morty' }

const result3 = getEnvVar( { ENV_VAR_1: true, ENV_VAR_2: true, ENV_VAR_3: true }, envVarSchema )
// const result3: {
//     ENV_VAR_1: string
//     ENV_VAR_2: string
//     ENV_VAR_3: string
// }
console.log( result3 ) // { ENV_VAR_1: 'rick', ENV_VAR_2: 'morty', ENV_VAR_3: 'mr meeseeks' }
```

**Reply by WesleyYue** [reply]
Thank you so much for your help. I really appreciate it, but that still doesn't quite get to what I am looking for. Your generic is a simpler version of what I have, but it also works the same way as what I have. I guess to illustrate the problem differently, I want to do this:

```ts
export function getEnvVar<
    Shape extends z.ZodRawShape,
    Mask extends { [ k in keyof Shape ]?: true },
> ( 
  mask: Mask, 
  schema: z.ZodObject<Shape> = envVarSchema  // Note the default value here
) {
    return schema.pick( mask ).parse( env )
}
```

but I can't because I get this error:
```ts
'{ ENV_VAR_1: z.ZodString; ENV_VAR_2: z.ZodString; ENV_VAR_3: z.ZodString; }' is assignable to the constraint of type 'Shape', but 'Shape' could be instantiated with a different subtype of constraint 'ZodRawShape'.
```

**Reply by WesleyYue** [reply]
And even if I change Shape to extend from EnvVarRaw and simplify the env to just one variable, I still get this non-obvious error:
`'{ ENV_1: z.ZodString; }' is assignable to the constraint of type 'Shape', but 'Shape' could be instantiated with a different subtype of constraint '{ ENV_1: ZodString; }'.ts(2322)`

**Comment by JacobWeisenburger** [comment, marked as answer]
Generics and default values don't play nice. I have had a lot of problems with this in the past and still don't fully understand it.

Would it work to just use `envVarSchema` inside the function instead of taking `schema` as an arg?

```ts
const envVarSchema = z.object( {
    ENV_VAR_1: z.string(),
    ENV_VAR_2: z.string(),
    ENV_VAR_3: z.string(),
} )

/* for node use: process.env */
/* for deno use: Deno.env.toObject() */
const env = {
    ENV_VAR_1: 'rick',
    ENV_VAR_2: 'morty',
    ENV_VAR_3: 'mr meeseeks',
}

export function getEnvVar<
    Schema extends typeof envVarSchema,
    Mask extends { [ k in keyof Schema[ 'shape' ] ]?: true },
> ( mask: Mask ) {
    return envVarSchema.pick( mask ).parse( env )
}

const result1 = getEnvVar( { ENV_VAR_1: true } )
// const result1: { ENV_VAR_1: string }
console.log( result1 ) // { ENV_VAR_1: 'rick' }

const result2 = getEnvVar( { ENV_VAR_2: true } )
// const result2: { ENV_VAR_2: string }
console.log( result2 ) // { ENV_VAR_2: 'morty' }

const result3 = getEnvVar( { ENV_VAR_1: true, ENV_VAR_2: true, ENV_VAR_3: true } )
// const result3: {
//     ENV_VAR_1: string
//     ENV_VAR_2: string
//     ENV_VAR_3: string
// }
console.log( result3 ) // { ENV_VAR_1: 'rick', ENV_VAR_2: 'morty', ENV_VAR_3: 'mr meeseeks' }
```

**Reply by WesleyYue** [reply]
Ah, yep, that's it! Seems like using `Schema[ 'shape' ]` is the key to making this work as it would otherwise have too many degrees of freedom. Thanks so much for your patience and help. I really appreciate it. Good to know that even experts find defaults + generics combo challenging.

**Reply by JacobWeisenburger** [reply]
lol, I'm glad you think I'm an expert. Thanks. :D
